src/software/networking/udp/proto_udp_listener.hpp

```diff
@@ -199,10 +240,33 @@ void ProtoUdpListener<ReceiveProtoT>::handleDataReception(
     }
 }
 
+template <class ReceiveProtoT>
+void ProtoUdpListener<ReceiveProtoT>::setupMulticast(
+    const boost::asio::ip::address& ip_address, const std::string& listen_interface,
+    std::optional<std::string>& error)
+{
+    if (ip_address.is_v4())
+    {
+        std::string interface_ip;
+        if (!getLocalIp(listen_interface, interface_ip))
+        {
+            std::stringstream ss;
+            ss << "Could not find the local ip address for the given interface: "
+               << listen_interface << std::endl;
+            error = ss.str();
+            return;
+        }
+        socket_.set_option(boost::asio::ip::multicast::join_group(
+            ip_address.to_v4(),
+            boost::asio::ip::address::from_string(interface_ip).to_v4()));
+        return;
+    }
+    socket_.set_option(boost::asio::ip::multicast::join_group(ip_address));
+}
+
 template <class ReceiveProtoT>
 ProtoUdpListener<ReceiveProtoT>::~ProtoUdpListener()
 {
-    close();
 }
```

Comment:
closing a socket twice results in a crash. This is usually not a problem because on `master`, this only happened at the end of thunderscope but when we change socket midgame, it takes out the python thread

Reply:
Where are we closing the socket the first time? Does its own destructor call close? Or are you referring to `RobotCommunication` manually calling close?

Reply:
https://github.com/UBC-Thunderbots/Software/blob/dea10f841ad24ef8ac8da743416e4c9495f9e83b/src/software/networking/udp/threaded_proto_udp_listener.hpp#L84

`ThreadedProtoUdpListener` calls close in the destructor and the `UdpListener`'s own destructor calls close as well causing this issue

Reply:
Oh interesting. What about the fact that we also call `.close` in `RobotCommunication.__exit__`? Should those also be removed?

Reply:
it's not the biggest problem since we're taking out the entirety of Thunderscope at that point but will remove